File: go/kdb/auth/resource_path.go
package auth

import "strings"

// ResourcePath is a structured view of a KDB resource address: database, optionally scoped
// down to a collection and a single document. Storage/transaction code keeps addressing
// resources with a flat namespaceId string ("database" or "database/collection"); this type
// exists only in the authorization layer, which needs to resolve grants at database, collection,
// and document granularity. See docs/kdb-rbac-plan.md.
type ResourcePath struct {
	Database   string
	Collection string // empty if not scoped to a collection
	DocumentID string // empty if not scoped to a document
}

func NewResourcePath(namespace string, documentID string) ResourcePath {
	if idx := strings.IndexByte(namespace, '/'); idx >= 0 {
		return ResourcePath{
			Database:   namespace[:idx],
			Collection: namespace[idx+1:],
			DocumentID: documentID,
		}
	}
	return ResourcePath{Database: namespace, DocumentID: documentID}
}

func (r ResourcePath) NamespaceID() string {
	if r.Collection == "" {
		return r.Database
	}
	return r.Database + "/" + r.Collection
}

// CandidatePaths returns grant-match candidates, most specific first: document, then
// collection, then database. A database-level grant covers every collection and document
// beneath it, and a collection-level grant covers every document in it.
func (r ResourcePath) CandidatePaths() []string {
	candidates := make([]string, 0, 3)
	if r.Collection != "" && r.DocumentID != "" {
		candidates = append(candidates, r.Database+"/"+r.Collection+"/"+r.DocumentID)
	}
	if r.Collection != "" {
		candidates = append(candidates, r.Database+"/"+r.Collection)
	}
	candidates = append(candidates, r.Database)
	return candidates
}

File: kdb-auth/src/commonMain/kotlin/dev/kdb/auth/ResourcePath.kt
package dev.kdb.auth

/**
 * Structured view of a KDB resource address: database, optionally scoped down to a collection
 * and a single document. Storage/transaction code keeps addressing resources with a flat
 * `namespaceId: String` (`"database"` or `"database/collection"`); this type exists only in the
 * authorization layer, which needs to resolve grants at database, collection, and document
 * granularity. See docs/kdb-rbac-plan.md.
 */
public data class ResourcePath(
    val database: String,
    val collection: String? = null,
    val documentId: String? = null,
) {
    public val namespaceId: String
        get() = if (collection != null) "$database/$collection" else database

    /**
     * Candidate grant-match paths, most specific first: document, then collection, then
     * database. Grant resolution checks these in order so a database-level grant covers every
     * collection and document beneath it, while a document-level grant can be looked up first
     * once deny grants are introduced.
     */
    public fun candidatePaths(): List<String> =
        buildList {
            if (collection != null && documentId != null) add("$database/$collection/$documentId")
            if (collection != null) add("$database/$collection")
            add(database)
        }

    public companion object {
        public fun of(
            namespaceId: String,
            documentId: String? = null,
        ): ResourcePath {
            val slash = namespaceId.indexOf('/')
            return if (slash < 0) {
                ResourcePath(database = namespaceId, collection = null, documentId = documentId)
            } else {
                ResourcePath(
                    database = namespaceId.substring(0, slash),
                    collection = namespaceId.substring(slash + 1),
                    documentId = documentId,
                )
            }
        }
    }
}
